File: client/src/router/connection.ts
import { Client } from "pg"

export class DbFactory {
    private static instance: DbFactory;
    private client: Client;
    private isConnected: boolean = false;

    private constructor() {
        this.client = new Client({
            host: process.env.POSTGRES_HOST,
            port: process.env.POSTGRES_PORT ? Number(process.env.POSTGRES_PORT) : 5432,
            database: process.env.POSTGRES_DB,
            user: process.env.POSTGRES_USER,
            password: process.env.POSTGRES_PASS
        })
    }
     
    public static getInstance(): DbFactory {
        if (!DbFactory.instance) {
            DbFactory.instance = new DbFactory();
        }
        return DbFactory.instance;
    }

    async connect() {
        try {
            if (!this.isConnected) {
                await this.client.connect();
                this.isConnected = true;
                console.log('Conectado a la base de datos PostgreSQL');
            }
        } catch (err) {
            console.error('Error conectando a la base de datos:', err);
            throw err;
        }
    }

    async disconnect() {
        try {
            if (this.isConnected) {
                await this.client.end();
                this.isConnected = false;
                console.log('🔌 Desconectado de la base de datos');
            }
        } catch (err) {
            console.error('❌ Error desconectando de la base de datos:', err);
            throw err;
        }
    }

    getClient(): Client {
        if (!this.isConnected) {
            throw new Error('La conexión a la base de datos no está establecida. Llama a connect() primero.');
        }
        return this.client;
    }

    isConnectedToDb(): boolean {
        return this.isConnected;
    }
}

File: server/src/db/connection.py
import psycopg
from dotenv import load_dotenv
import os
load_dotenv()
class DbFactory:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance.client = None
            cls._instance.is_connected = False
        return cls._instance

    def __init__(self):
        self.config = {
            'host': os.getenv('POSTGRES_HOST'),
            'port': int(os.getenv('POSTGRES_PORT', 5432)),
            'dbname': os.getenv('POSTGRES_DB'),
            'user': os.getenv('POSTGRES_USER'),
            'password': os.getenv('POSTGRES_PASSWORD'),
        }

    def connect(self):
        if not self.is_connected:
            self.client = psycopg.connect(**self.config)
            self.is_connected = True
            print("Conectado a la base de datos PostgreSQL")

    def disconnect(self):
        if self.is_connected and self.client:
            self.client.close()
            self.is_connected = False
            print("Desconectado de la base de datos")

    def get_client(self):
        if not self.is_connected:
            raise RuntimeError("La conexión a la base de datos no está establecida. Llama a connect() primero.")
        return self.client

    def is_connected_to_db(self):
        return self.is_connected